plugin/aws/s3Bucket.go

```diff
@@ -198,7 +207,33 @@ func (b *s3Bucket) Metadata(ctx context.Context) (plugin.MetadataMap, error) {
 		return nil, err
 	}
 
-	metadata["region"] = b.region
+	region, err := b.getRegion(ctx)
+	if err != nil {
+		return nil, err
+	}
+	metadata["region"] = region
 
 	return metadata, nil
 }
+
+func (b *s3Bucket) getRegion(ctx context.Context) (string, error) {
+	var err error
+	b.regioner.Do(func() {
+		locRequest := &s3Client.GetBucketLocationInput{Bucket: awsSDK.String(b.Name())}
+		resp, reqErr := b.client.GetBucketLocationWithContext(ctx, locRequest)
+		if reqErr != nil {
+			// TODO: Should we log a warning and continue instead of returning an error?
+			err = fmt.Errorf("could not get the region of bucket %v: %v", b.Name(), reqErr)
+		}
+
+		// The response will be empty if the bucket is in Amazon's default region (us-east-1)
+		b.region = s3Client.NormalizeBucketLocation(awsSDK.StringValue(resp.LocationConstraint))
+		// Update client to be region-specific
+		b.client = s3Client.New(b.session, aws.NewConfig().WithRegion(b.region))
+	})
+	if err != nil {
```

Comment:
Oh yeah, this code will only get executed by a single thread b/c err's set in Once which also gets executed by that same thread. A comment describing that would help because it requires a bit of thinking.